### PR TITLE
Use optimized folly DistributedMutex in LRUCache when available

### PR DESCRIPTION
**Pull request by pdillinger**
Summary: folly DistributedMutex is faster than standard mutexes though
imposes some static obligations on usage. See
https://github.com/facebook/folly/blob/main/folly/synchronization/DistributedMutex.h
for details. Here we use this alternative for our Cache implementations
(especially LRUCache) for better locking performance, when RocksDB is
compiled with folly.

Also added information about which distributed mutex implementation is
being used to cache_bench output and to DB LOG.

Intended follow-up:
* Use DMutex in more places, perhaps improving API to support non-scoped
locking
* Fix linking with fbcode compiler (needs ROCKSDB_NO_FBCODE=1 currently)

Credit: Thanks Siying for reminding me about this line of work that was previously
left unfinished.

Test Plan: for correctness, existing tests. CircleCI config updated.
Also Meta-internal buck build updated. 

For performance, ran simultaneous before & after cache_bench. Out of three
comparison runs, the middle improvement to ops/sec was +21%:

Baseline: USE_CLANG=1 DEBUG_LEVEL=0 make -j24 cache_bench (fbcode
compiler)

```
Complete in 20.201 s; Rough parallel ops/sec = 1584062
Thread ops/sec = 107176

Operation latency (ns):
Count: 32000000 Average: 9257.9421  StdDev: 122412.04
Min: 134  Median: 3623.0493  Max: 56918500
Percentiles: P50: 3623.05 P75: 10288.02 P99: 30219.35 P99.9: 683522.04 P99.99: 7302791.63
```

New: (add USE_FOLLY=1)

```
Complete in 16.674 s; Rough parallel ops/sec = 1919135  (+21%)
Thread ops/sec = 135487

Operation latency (ns):
Count: 32000000 Average: 7304.9294  StdDev: 108530.28
Min: 132  Median: 3777.6012  Max: 91030902
Percentiles: P50: 3777.60 P75: 10169.89 P99: 24504.51 P99.9: 59721.59 P99.99: 1861151.83
```

And with db_bench (both `make -j24 DEBUG_LEVEL=0 ROCKSDB_NO_FBCODE=1 USE_FOLLY=1 db_bench` with same folly checkout)
Create db with
```
TEST_TMPDIR=/dev/shm/rocksdb_bench ./db_bench -benchmarks=fillrandom -num=10000000 -disable_wal=1 -write_buffer_size=30000000 -bloom_bits=16 -compaction_style=2 -fifo_compaction_max_table_files_size_mb=10000 -fifo_compaction_allow_compaction=0
```

Single-threaded read test with
```
TEST_TMPDIR=/dev/shm/rocksdb_bench ./db_bench -benchmarks=readrandom[-X1000] -num=10000000 -readonly -bloom_bits=16 -compaction_style=2 -fifo_compaction_max_table_files_size_mb=10000 -fifo_compaction_allow_compaction=0 -cache_index_and_filter_blocks=1 -cache_size=1000000000 -duration 20
```
suggests a small regression (-3.7%) with single thread (before and after run at the same time):

Before: readrandom [AVG 23 runs] : 42644 (± 1121) ops/sec;    3.0 (± 0.1) MB/sec
After: readrandom [AVG 23 runs] : 41048 (± 1063) ops/sec;    2.9 (± 0.1) MB/sec

But adding -threads=8 shows big benefit (+12.9%):

Before: readrandom [AVG 8 runs] : 164895 (± 6322) ops/sec;   11.5 (± 0.4) MB/sec
After: readrandom [AVG 8 runs] : 186186 (± 6761) ops/sec;   13.0 (± 0.5) MB/sec

And -threads=4 a smaller benefit (+4.7%):

Before: readrandom [AVG 30 runs] : 96933 (± 3375) ops/sec;    6.8 (± 0.2) MB/sec
After: readrandom [AVG 30 runs] : 101471 (± 2794) ops/sec;    7.1 (± 0.2) MB/sec

Nevertheless, I think this is overall a good change for most users because it improves performance under realistic "stress" conditions.

### CODE REVIEW

**Comment by mrambacher** [comment]
Curious as to the reason for the new DMutex class.  Why not change port::Mutex and related classes instead, then everything will get the folly mutex if so desired.  The existing port::Mutex could be port::Posix/WindowsMutex and the DMutex could be port::Mutex.

That seems like a cleaner solution that all tools can take advantage of, rather than this solution requiring many more changes.

**Comment by pdillinger** [comment]
> Curious as to the reason for the new DMutex class. Why not change port::Mutex and related classes instead, then everything will get the folly mutex if so desired. The existing port::Mutex could be port::Posix/WindowsMutex and the DMutex could be port::Mutex.
> 
> That seems like a cleaner solution that all tools can take advantage of, rather than this solution requiring many more changes.

Because it would be a time-consuming change to make. https://github.com/facebook/folly/blob/e9ff81f299c4adaed5f2fec0e0a90334c15f9f0e/folly/synchronization/DistributedMutex.h#L103-L105

**Review by anand1976 — COMMENTED** [review]
Can we also run db_bench to check for any regression?

File: Makefile [hunk 1]
```diff
@@ -2357,10 +2359,14 @@ checkout_folly:
 	fi
 	@# Pin to a particular version for public CI, so that PR authors don't
 	@# need to worry about folly breaking our integration. Update periodically
-	cd third-party/folly && git reset --hard 98b9b2c1124e99f50f9085ddee74ce32afffc665
+	cd third-party/folly && git reset --hard beacd86d63cd71c904632262e6c36f60874d78ba
```
[the remark this answers is not in the document]
Comment (pdillinger) [reply Makefile:2362]: Just opportunistic 👍

**Comment by pdillinger** [comment]
Sure I can also run db_bench.

**Review by anand1976 — APPROVED** [review]
Thanks for the db_bench results. The multi-thread results look good. The distributed mutex documentation does mention that its slower for the uncontended case, so I guess the single thread regression is expected.

**Pull request closed without merging**